internal/modelcontroller/pods.go

```diff
@@ -0,0 +1,131 @@
+package modelcontroller
+
+import (
+	"context"
+	"fmt"
+
+	kubeaiv1 "github.com/substratusai/kubeai/api/v1"
+	"github.com/substratusai/kubeai/internal/k8sutils"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func (r *ModelReconciler) calculatePodPlan(allPods *corev1.PodList, model *kubeaiv1.Model, modelConfig ModelConfig) *podPlan {
+	var desiredReplicas int32
+	// NOTE: Replicas could be nil if autoscaling is disabled.
+	if model.Spec.Replicas != nil {
+		desiredReplicas = *model.Spec.Replicas
+	}
+
+	var podForModel func(*kubeaiv1.Model, ModelConfig, string) *corev1.Pod
+	switch model.Spec.Engine {
+	case kubeaiv1.OLlamaEngine:
+		podForModel = r.oLlamaPodForModel
+	case kubeaiv1.FasterWhisperEngine:
+		podForModel = r.fasterWhisperPodForModel
+	default:
+		podForModel = r.vLLMPodForModel
+	}
+
+	var (
+		toCreate []*corev1.Pod
+		toDelete []*corev1.Pod
+	)
+
+	podMap := make(map[string]corev1.Pod, len(allPods.Items))
+	for _, pod := range allPods.Items {
+		podMap[pod.Name] = pod
+	}
+
+	// Loop through a deterministic list of Pod names and create or delete Pods as needed.
+	// Because the client used to list Pods is a cache client, the exact number of Pods
+	// returned may not be accurate which is why we don't just compare total count of Pods
+	// to desired replicas.
+	for i := int32(0); i < desiredReplicas; i++ {
+		name := fmt.Sprintf("model-%s-%d", model.Name, i)
+
+		expectedPod := podForModel(model, modelConfig, name)
+		// TODO: If collisions become an issue, we can add a Model.Status.CollisionCount (new field)
+		// to the PodHash call.
+		expectedPodHash := k8sutils.PodHash(expectedPod.Spec, nil)
+		k8sutils.SetLabel(expectedPod, kubeaiv1.PodHashLabel, expectedPodHash)
+
+		currentPod, ok := podMap[name]
+		if ok {
+			// Already exists, compare to see if it needs to be recreated.
+			// TODO: Consider implementing a rollout strategy.
+			// Right now, we just delete and recreate all Pods.
```

Comment:
This could be problematic for online serving of larger models. We shouldn't release this imo until we have a rollout strategy that ensures there is always at least x ready pods.

Reply:
I am thinking of something like maxUnreadyPods